### PR TITLE
[SPARK][VARIANT] Add minimal support for variant type in delta-spark

### PR DESCRIPTION
**Pull request by richardc-db**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Adds the variant table feature to minimally implement the variant type as described in the RFC in #2867.

Also disables using variant columns as partition columns.

## How was this patch tested?

Added some UTs. More UTs will be merged in followup PRs

tested against both spark 3.5 and 4.0 snapshot with
```
build/sbt -DsparkVersion=latest spark/'testOnly org.apache.spark.sql.delta.DeltaVariantSuite'
build/sbt -DsparkVersion=master spark/'testOnly org.apache.spark.sql.delta.DeltaVariantSuite'
```

## Does this PR introduce _any_ user-facing changes?

no

### CODE REVIEW

File: spark/src/test/scala-spark-master/org/apache/spark/sql/delta/DeltaVariantSuite.scala [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import org.apache.spark.sql.{AnalysisException, QueryTest, Row}
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types.StructType
+
+class DeltaVariantSuite
+    extends QueryTest
+    with SharedSparkSession
+    with DeltaSQLCommandTest {
+
```
Comment (scottsand-db) [inline spark/src/test/scala-spark-master/org/apache/spark/sql/delta/DeltaVariantSuite.scala:33]: nit: extra line

File: spark/src/main/scala/org/apache/spark/sql/delta/TableFeature.scala [hunk 2]
```diff
@@ -502,6 +503,14 @@ object TimestampNTZTableFeature extends ReaderWriterFeature(name = "timestampNtz
   }
 }
 
+object VariantTypeTableFeature extends ReaderWriterFeature(name = "variantType")
```
Comment (scovich) [inline spark/src/main/scala/org/apache/spark/sql/delta/TableFeature.scala:506]:
```suggestion
object VariantTypeTableFeature extends ReaderWriterFeature(name = "variantType-dev")
```
(the protocol RFC has not been accepted yet)

File: spark/src/main/scala-spark-master/shims/VariantShims.scala [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.types
+
+object VariantShims {
+
+  /** Spark's variant type is only implemented in Spark 4.0 and above. */
+  def isTypeVariant(dt: DataType): Boolean = dt.isInstanceOf[VariantType]
```
Comment (scovich) [inline spark/src/main/scala-spark-master/shims/VariantShims.scala:22]: nit
```suggestion
  def isVariantType(dt: DataType): Boolean = dt.isInstanceOf[VariantType]
```

File: spark/src/main/scala/org/apache/spark/sql/delta/TableFeature.scala (hunk 2 above)
Comment (richardc-db) [reply spark/src/main/scala/org/apache/spark/sql/delta/TableFeature.scala]: added the "-dev" suffix back

**Review by scottsand-db — APPROVED** [review]

**Comment by scottsand-db** [comment]
Spark master failures are unrelated

**Pull request merged**